src/runtime/runtime.ts

```diff
@@ -784,6 +789,31 @@ interface Node {
   exec(index:Index, input:Change, prefix:ID[], transaction:number, round:number, results?:ID[][], changes?:Change[]):boolean;
 }
 
+/**
+ * The JoinNode implements generic join across multiple constraints.
+ * Since our system is incremental, we need to do something slightly fancier than we did in the previous runtime.
+ * For each new change that enters the system, we ask each of our constraints whether they are capable of producing
+ * a new result. In the case where a single constraint can, we presolve that constraint and then run the rest normally,
+ * limited to only producing results that match the first constraint. However, if multiple constraints might apply the input,
+ * we need to run for each *combination* of heads. E.g.:
+ *
+ * Given a join node with constraints [A, B, C, and D], where A and C can both apply the input, we must combine the results of
+ * the following computations to get the full result set:
+ * Apply A -> Do {B, C, D}
+ * Apply {A, D} -> Do {B, C}
+ * Apply {D} -> Do {A, B, C}
+ *
+ * We calculate this using the power set in exec.
+ *
+ * We then apply each of these combinations by running a genericJoin over the remaining unresolved registers.
```

Comment:
the comment lines here are all pretty long, probably should flow them onto another line so no scrolling is necessary to read.